omnigibson/transition_rules.py

```diff
@@ -296,358 +710,453 @@ def transition(self, individual_objects, group_objects):
-        # For every object in group_objects, we remove them from the simulator
-        for i, (obj_category, objs) in enumerate(group_objects.items()):
-            for j, obj in enumerate(objs):
-                t_results.remove.append(obj)
-
-        # Remove all physical particles that are inside the blender
-        for system_name in self.particle_requirements.keys():
-            system = get_system(system_name)
-            # No need to check for whether particle instancers exist because they must due to @self.condition passing!
-            for inst in system.particle_instancers.values():
-                indices = self._check_in_volume[blender.name](inst.particle_positions).nonzero()[0]
-                inst.remove_particles(idxs=indices)
-
-        # Spawn in blended physical particles!
-        blender.states[Filled].set_value(get_system(self.output_system), True)
+    @classmethod
+    def initialize(cls):
+        # Cache active recipes given the current set of objects
+        cls._ACTIVE_RECIPES = dict()
+        cls._CATEGORY_IDXS = dict()
+        cls._OBJECTS = []
+
+        # Prune any recipes whose objects / system requirements are not met by the current set of objects / systems
+        objects_by_category = og.sim.scene.object_registry.get_dict("category")
+
+        for name, recipe in cls._RECIPES.items():
+            # Check valid active systems
+            if not cls._validate_recipe_systems_exist(recipe=recipe):
+                continue
+
+            # Check valid object quantities
+            if not cls._validate_recipe_objects_exist(recipe=recipe):
+                continue
+
+            # All pre-requisites met, add to active recipes
+            cls._ACTIVE_RECIPES[name] = recipe
+
+        # Finally, compute relevant objects and category mapping based on relevant categories
+        i = 0
+        for category, objects in objects_by_category.items():
+            cls._CATEGORY_IDXS[category] = i + np.arange(len(objects))
+            cls._OBJECTS += list(objects)
+            i += len(objects)
+
+        # Wrap relevant objects as numpy array so we can index into it efficiently
+        cls._OBJECTS = np.array(cls._OBJECTS)
+
+    @classproperty
+    def group_filters(cls):
+        # Fillable object required
+        return {"container": AbilityFilter(ability="fillable")}
 
-        return t_results
+    @classmethod
+    def transition(cls, individual_objects, group_objects):
+        t_results = TransitionResults()
 
+        # Compute all relevant object AABB positions
+        obj_positions = np.array([obj.aabb_center for obj in cls._OBJECTS])
+
+        # Iterate over all fillable objects, to execute recipes for each one
+        for container in individual_objects["container"]:
+            # Compute in volume for all relevant object positions
+            contained_particles_state = container.states[ContainedParticles]
+            in_volume = contained_particles_state.check_in_volume(obj_positions)
```

Comment:
Interesting, I guess this is far more efficient than checking `obj.states[Inside].get_value(container)` for each object.